pkg/operator/controller/controller_dns.go

```diff
@@ -1,89 +1,139 @@
+		if err := r.client.Create(context.TODO(), desired); err != nil {
+			return nil, fmt.Errorf("failed to create dnsrecord %s/%s: %v", desired.Namespace, desired.Name, err)
+		}
+		log.Info("created dnsrecord", "namespace", desired.Namespace, "dnsrecord", desired)
+		return desired, nil
 	}
+	return current, nil
 }
 
-// desiredDNSRecords will return any necessary DNS records for the given inputs.
-// If an ingress domain is in use, records are desired in every specified zone
-// present in the cluster DNS configuration.
-func desiredDNSRecords(ci *operatorv1.IngressController, dnsConfig *configv1.DNS, service *corev1.Service) []*dns.Record {
-	records := []*dns.Record{}
-
+// ensureWildcardDNSRecord will return any necessary wildcard DNS records for the
+// ingresscontroller.
+//
+// For now, if the service has more than one .status.loadbalancer.ingress, only
+// the first will be used.
+//
+// TODO: If .status.loadbalancer.ingress is processed once as non-empty and then
+// later becomes empty, what should we do? Currently we'll treat it as an intent
+// to not have a desired record.
+func desiredWildcardRecord(ic *operatorv1.IngressController, service *corev1.Service) *iov1.DNSRecord {
 	// If the ingresscontroller has no ingress domain, we cannot configure any
 	// DNS records.
-	if len(ci.Status.Domain) == 0 {
-		return records
+	if len(ic.Status.Domain) == 0 {
+		return nil
 	}
 
-	// If the HA type is not cloud, then we don't manage DNS.
-	if ci.Status.EndpointPublishingStrategy.Type != operatorv1.LoadBalancerServiceStrategyType {
-		return records
+	// DNS is only managed for LB publishing.
+	if ic.Status.EndpointPublishingStrategy.Type != operatorv1.LoadBalancerServiceStrategyType {
+		return nil
 	}
 
-	name := fmt.Sprintf("*.%s", ci.Status.Domain)
-	zones := []configv1.DNSZone{}
-	if dnsConfig.Spec.PrivateZone != nil {
-		zones = append(zones, *dnsConfig.Spec.PrivateZone)
+	// No LB target exists for the domain record to point at.
+	if len(service.Status.LoadBalancer.Ingress) == 0 {
+		return nil
 	}
-	if dnsConfig.Spec.PublicZone != nil {
-		zones = append(zones, *dnsConfig.Spec.PublicZone)
+
+	ingress := service.Status.LoadBalancer.Ingress[0]
+
+	// Quick sanity check since we don't know how to handle both being set (is
+	// that even a valid state?)
+	if len(ingress.Hostname) > 0 && len(ingress.IP) > 0 {
+		return nil
 	}
-	for _, ingress := range service.Status.LoadBalancer.Ingress {
-		if len(ingress.Hostname) > 0 {
-			for _, zone := range zones {
-				records = append(records, newAliasRecord(name, ingress.Hostname, zone))
-			}
-		}
-		if len(ingress.IP) > 0 {
-			for _, zone := range zones {
-				records = append(records, newARecord(name, ingress.IP, zone))
-			}
+
+	name := WildcardDNSRecordName(ic)
+	domain := fmt.Sprintf("*.%s", ic.Status.Domain)
+	var target string
+	var recordType string
+
+	if len(ingress.Hostname) > 0 {
+		recordType = iov1.CNAMERecordType
+		target = ingress.Hostname
+	} else {
```

Comment:
Could the service have `.Status.LoadBalancer.Ingress[0]` with empty `.Hostname` and `.IP`?

Reply:
Not sure, there are a lot of unresolved Q's about that status field and our handling of them... but the environments we support are predictable enough to support our current assumptions, so I'm not really focused on solving them in this PR